pkg/test/golden/validator.go

```diff
@@ -0,0 +1,329 @@
+			t.Errorf("expected exactly one item in %s", p)
+			continue
+		}
+
+		crJSON, err := objs.Items[0].JSON()
+		if err != nil {
+			t.Errorf("error converting CR to json in %s: %v", p, err)
+			continue
+		}
+
+		cr, _, err := serializer.Decode(crJSON, nil, nil)
+		if err != nil {
+			t.Errorf("error parsing CR in %s: %v", p, err)
+			continue
+		}
+
+		namespace, err := metadataAccessor.Namespace(cr)
+		if err != nil {
+			t.Errorf("error getting namespace in %s: %v", p, err)
+			continue
+		}
+
+		name, err := metadataAccessor.Name(cr)
+		if err != nil {
+			t.Errorf("error getting name in %s: %v", p, err)
+			continue
+		}
+
+		nsn := types.NamespacedName{Namespace: namespace, Name: name}
+
+		objects, err := r.BuildDeploymentObjects(ctx, nsn, cr.(declarative.DeclarativeObject))
+		if err != nil {
+			t.Errorf("error building deployment objects: %v", err)
+			continue
+		}
+
+		var actualYAML string
+		{
+			var b bytes.Buffer
+
+			for i, o := range objects.Items {
+				if i != 0 {
+					b.WriteString("\n---\n\n")
+				}
+				u := o.UnstructuredObject()
+				if err := yamlizer.Encode(u, &b); err != nil {
+					t.Fatalf("error encoding to yaml: %v", err)
+				}
+			}
+
+			actualYAML = b.String()
+		}
+
+		expectedPath := strings.Replace(p, ".in.yaml", ".out.yaml", -1)
+		var expectedYAML string
+		{
+			b, err := ioutil.ReadFile(expectedPath)
+			if err != nil {
+				t.Errorf("error reading file %s: %v", expectedPath, err)
+				continue
+			}
+			expectedYAML = string(b)
+		}
+
+		if actualYAML != expectedYAML {
+			if os.Getenv("HACK_AUTOFIX_EXPECTED_OUTPUT") != "" {
+				t.Logf("HACK_AUTOFIX_EXPECTED_OUTPUT is set; replacing expected output in %s", expectedPath)
+				if err := ioutil.WriteFile(expectedPath, []byte(actualYAML), 0644); err != nil {
+					t.Fatalf("error writing expected output to %s: %v", expectedPath, err)
+				}
+				continue
+			}
+
+			if err := diffFiles(t, expectedPath, actualYAML); err != nil {
+				t.Logf("failed to run system diff, falling back to string diff: %v", err)
+				t.Logf("diff: %s", diff.StringDiff(actualYAML, expectedYAML))
+			}
+
+			t.Errorf("unexpected diff between actual and expected YAML. See previous output for details.")
+			t.Logf(`To regenerate the output based on this result, rerun this test with HACK_AUTOFIX_EXPECTED_OUTPUT="true"`)
+		}
+
+	}
+}
+
+func diffFiles(t *testing.T, expectedPath, actual string) error {
+	t.Helper()
+	writeTmp := func(content string) (string, error) {
+		tmp, err := ioutil.TempFile("", "*.yaml")
+		if err != nil {
+			return "", err
+		}
+		defer func() {
+			tmp.Close()
+		}()
+		if _, err := tmp.Write([]byte(content)); err != nil {
+			return "", err
+		}
+		return tmp.Name(), nil
+	}
+
+	actualTmp, err := writeTmp(actual)
+	if err != nil {
+		return errors.Wrapf(err, "write actual yaml to temp file failed")
+	}
+	t.Logf("Wrote actual to %s", actualTmp)
+
+	// pls to use unified diffs, kthxbai?
+	cmd := exec.Command("diff", "-u", expectedPath, actualTmp)
```

Comment:
I think k8s has a diff package, but I think it's a great idea to rely on diff instead as the output from the k8s diff is really awkward, and kops wrote our own diff package and it's a massive pain...